test/solver_tests/spectral_test.py

```diff
@@ -10,157 +10,149 @@
-    #         mutation_frequencies,
-    #         -1,
-    #         list(spsolver.unique_character_matrix.index),
-    #         similarity_function=dissimilarity_functions.hamming_similarity_without_missing,
-    #         w=weights,
-    #     )
-
-    #     self.assertEqual(G["c1"]["c3"]["weight"], 1)
-    #     self.assertEqual(G["c3"]["c4"]["weight"], 3)
-    #     self.assertEqual(G["c4"]["c5"]["weight"], 1)
-    #     self.assertNotIn(["c1", "c4"], G.edges)
-    #     self.assertNotIn(["c1", "c5"], G.edges)
-    #     self.assertNotIn(["c2", "c5"], G.edges)
-
-    # def test_hill_climb(self):
-    #     G = nx.Graph()
-    #     for i in range(4):
-    #         G.add_node(i)
-    #     G.add_edge(0, 1, weight=3)
-    #     G.add_edge(0, 2, weight=2)
-    #     G.add_edge(0, 3, weight=2)
-    #     G.add_edge(1, 2, weight=2)
-    #     G.add_edge(1, 3, weight=10)
-    #     G.add_edge(2, 3, weight=1)
-
-    #     new_cut = graph_utilities.spectral_improve_cut(G, [0])
-    #     self.assertEqual(new_cut, [0, 2])
+    def test_similarity(self):
+        cm = pd.DataFrame.from_dict(
+            {
+                "c1": [5, -1, 1, 2, 0],
+                "c2": [5, 4, 0, 2, 1],
+                "c3": [4, 4, 3, 1, 1],
+                "c4": [-1, 4, 0, 1, -1],
+            },
+            orient="index",
+            columns=["a", "b", "c", "d", "e"],
+        )
+
+        self.assertEqual(
+            dissimilarity_functions.hamming_similarity_without_missing(
+                list(cm.loc["c1", :]), list(cm.loc["c3", :]), -1
+            ),
+            0,
+        )
+        self.assertEqual(
+            dissimilarity_functions.hamming_similarity_without_missing(
+                list(cm.loc["c2", :]), list(cm.loc["c3", :]), -1
+            ),
+            2,
+        )
+        self.assertEqual(
+            dissimilarity_functions.hamming_similarity_without_missing(
+                list(cm.loc["c3", :]), list(cm.loc["c4", :]), -1
+            ),
+            2,
+        )
+
+    def test_similarity_weighted(self):
+        cm = pd.DataFrame.from_dict(
+            {
+                "c1": [5, -1, 1, 2, 0],
+                "c2": [5, 4, 0, 2, 1],
+                "c3": [4, 4, 3, 1, 1],
+                "c4": [-1, 4, 0, 1, -1],
+            },
+            orient="index",
+            columns=["a", "b", "c", "d", "e"],
+        )
+
+        weights = {
+            0: {5: 2},
+            1: {4: 2},
+            2: {1: 1, 3: 1},
+            3: {1: 1, 2: 1},
+            4: {1: 3},
+        }
+
+        self.assertEqual(
+            dissimilarity_functions.hamming_similarity_without_missing(
+                list(cm.loc["c1", :]), list(cm.loc["c3", :]), -1, weights
+            ),
+            0,
+        )
+        self.assertEqual(
+            dissimilarity_functions.hamming_similarity_without_missing(
+                list(cm.loc["c2", :]), list(cm.loc["c3", :]), -1, weights
+            ),
+            5,
+        )
```

Comment:
Lots of potential for sporadic behavior due to randomness in the weights, potentially. Take a look at the comments above with suggestions for dealing with this.